### PR TITLE
Review current state of amidlash/grpc "kokoro_experiment"

### PR DESCRIPTION
**Pull request by jtattermusch**
I'm not planning to submit this PR, I'm simply providing feedback on the code that's currently in the  https://github.com/amidlash/grpc/tree/kokoro_experiment branch.

@amidlash

### CODE REVIEW

File: tools/internal_ci/linux/grpc_aws_experiment.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+IDENTITY=${KOKORO_KEYSTORE_DIR}/73836_grpc_arm_instance_ssh_private_test_key1
+
+if [ -z "$INSTANCE" ]; then
+    echo "no INSTANCE set, aborting"
+    exit 1
+fi
+
+FILE=grpc_aws_experiment_remote.sh
+chmod 700 $IDENTITY
+ssh -i $IDENTITY -o StrictHostKeyChecking=no ubuntu@$INSTANCE "rm -rf grpc"  # TODO: remove when using fresh instances
```
Comment (jtattermusch) [inline tools/internal_ci/linux/grpc_aws_experiment.sh:21]: nit: I think it would be handy to put all the files in a "workspace" subdirectory, just for separation.
e.g the script could create "~/remote_executor_workspace" directory, which would make it easier to keep all the files contained and also to support other codebases than just grpc.

**Review by jtattermusch — COMMENTED** [review]
Initial set of comments.

File: tools/internal_ci/linux/grpc_aws_experiment.sh (hunk 1 above)
Comment (amidlash) [reply tools/internal_ci/linux/grpc_aws_experiment.sh:21]: I can create another directory if you'd like. Ultimately, however, each vm will be used once. In this case, I'm not sure what the difference is. Would we ever check out multiple codebases in a single test run?

Comment (jtattermusch) [reply tools/internal_ci/linux/grpc_aws_experiment.sh:21]: I'd prefer to have that directory. I agree that currently it's not strictly required (since we are going to be throwing away the VMs), but in the future it might come handy. It might also make logs more digestible since the purpose of that directory is obvious just from its name.

File: tools/internal_ci/linux/grpc_aws_experiment.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+fi
+
+IDENTITY=${KOKORO_KEYSTORE_DIR}/73836_grpc_arm_instance_ssh_private_test_key1
+
+if [ -z "$INSTANCE" ]; then
+    echo "no INSTANCE set, aborting"
+    exit 1
+fi
+
+FILE=grpc_aws_experiment_remote.sh
```
[the remark this answers is not in the document]
Comment (amidlash) [reply tools/internal_ci/linux/grpc_aws_experiment.sh:19]: Sure. Though honestly I structured things this way to avoid the risk of typo , given limited static checking in bash.

How about REMOTE_SCRIPT as a compromise?

File: tools/internal_ci/linux/grpc_aws_experiment_remote.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
```
[the remark this answers is not in the document]
Comment (amidlash) [reply tools/internal_ci/linux/grpc_aws_experiment_remote.sh:2]: Sounds good, I'll integrate this in the next commit.

File: tools/internal_ci/linux/grpc_aws_experiment.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
```
[the remark this answers is not in the document]
Comment (amidlash) [reply tools/internal_ci/linux/grpc_aws_experiment.sh:2]: Included in next commit.

File: tools/internal_ci/linux/grpc_aws_experiment.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+scp -i $IDENTITY -o StrictHostKeyChecking=no -r github/grpc ubuntu@$INSTANCE:
+ssh -i $IDENTITY -o StrictHostKeyChecking=no ubuntu@$INSTANCE "uname -a; ls -l; bash grpc/tools/internal_ci/linux/$FILE" > ssh_log.txt
+cat ssh_log.txt
+# Sync back sponge_log artifacts (wip)
+echo "looking for sponge logs..."
+find . | grep sponge_log
+
+
+# Match return value
+RETURN=$(tail -n1 ssh_log.txt | sed "s/RETURN=//")
```
[the remark this answers is not in the document]
Comment (amidlash) [reply tools/internal_ci/linux/grpc_aws_experiment.sh:31]: Ah, I didn't realize the ssh command mirrors the output. This is much simpler.

Storing the log was only necessary to capture RETURN, I'll remove this.

File: tools/internal_ci/linux/grpc_aws_experiment_remote.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+git submodule update --init # 1m
+# build with cmake
+mkdir -p cmake/build
+cd cmake/build
+cmake ../..
+make
+
+# run tests
+cd ../../
+python tools/run_tests/run_tests.py -l c++
```
[the remark this answers is not in the document]
Comment (amidlash) [reply tools/internal_ci/linux/grpc_aws_experiment_remote.sh:21]: Sure, there was just some extra complexity around getting bazel to run on ARM. It looks like we'll need to build bazel, which requires building from source.

[the remark this answers is not in the document]
Comment (jtattermusch) [reply tools/internal_ci/linux/grpc_aws_experiment_remote.sh:21]: Actually, it worked just fine for me.
When executing the tests in the grpc/grpc repo, I'm actually not running just the `bazel` command, but I'm running `tools/bazel` script (see here: https://github.com/grpc/grpc/blob/master/tools/bazel).
The shell wrapper is there to make sure to always use the "right" version of bazel (grpc repo as a whole uses a single version of bazel for all the testing for a better and more uniform experience). The script will download the right arm64 version of bazel for you.

So basically, once the prerequisites are installed (I think it's `clang` in addition to what you need for cmake build, see the manual testing doc), you can just run `tools/bazel  test  foo/bar` and it magically works (and downloads bazel automatically).

File: tools/internal_ci/linux/grpc_aws_experiment.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+FILE=grpc_aws_experiment_remote.sh
+chmod 700 $IDENTITY
+ssh -i $IDENTITY -o StrictHostKeyChecking=no ubuntu@$INSTANCE "rm -rf grpc"  # TODO: remove when using fresh instances
+scp -i $IDENTITY -o StrictHostKeyChecking=no -r github/grpc ubuntu@$INSTANCE:
+ssh -i $IDENTITY -o StrictHostKeyChecking=no ubuntu@$INSTANCE "uname -a; ls -l; bash grpc/tools/internal_ci/linux/$FILE" > ssh_log.txt
+cat ssh_log.txt
+# Sync back sponge_log artifacts (wip)
+echo "looking for sponge logs..."
+find . | grep sponge_log
```
[the remark this answers is not in the document]
Comment (jtattermusch) [reply tools/internal_ci/linux/grpc_aws_experiment.sh]: They are generated by run_tests.py, but there's a cmdline argument to enable the xml reports.

On our CI, this is the command that our continuous C/C++ runs executes:
```
Will run these tests:
  run_tests_c_linux_opt_native: "python tools/run_tests/run_tests.py --use_docker -t -j 16 -x run_tests/c_linux_opt_native/sponge_log.xml --report_suite_name c_linux_opt_native -l c -c opt --iomgr_platform native --bq_result_table aggregate_results --measure_cpu_costs"
  run_tests_c++_linux_opt_native: "python tools/run_tests/run_tests.py --use_docker -t -j 16 -x run_tests/cpp_linux_opt_native/sponge_log.xml --report_suite_name cpp_linux_opt_native -l c++ -c opt --iomgr_platform native --bq_result_table aggregate_results --measure_cpu_costs"
```
(see log https://source.cloud.google.com/results/invocations/22e4ad26-a871-4cc2-a5a5-383cd958812b/log)

The `-x run_tests/c_linux_opt_native/sponge_log.xml` is what enables the sponge_log.xml
The arguments `--use_docker`, `--iomgr_platform native`, `--bq_result_table aggregate_results` and `--measure_cpu_costs`) you can ignore. For `--use_docker` docker you'd need a ARM64 version of the base docker image first ; `--iomgr_platform native` is the default;  `--bq_result_table aggregate_results` wouldn't work since the AWS instance doesn't have credentials to upload to bigquery. 

Note that the bazel run doesn't generate a sponge_log.xml so the best you can do for bazel now is to store the bazel log. (bazel actually can upload results to resultstore via the build events protocol, but you'd need credentials for that and they won't be available on the AWS machine).

File: tools/internal_ci/linux/grpc_aws_experiment.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtattermusch) [reply tools/internal_ci/linux/grpc_aws_experiment.sh:19]: Sg.

**Comment by jtattermusch** [comment]
@amidlash I wasn't sure if the new commits were something for me to review  or if you just pushed commits to your branch to test them.
Please let me know once there's something to review.
If this PR is not useful for feedback anymore, we can close it.

**Comment by jtattermusch** [comment]
I'm going to close this PR as it's pointed to a wrong branch.

@amidlash  once you have something that you want me to review, please create a new PR against this branch: https://github.com/jtattermusch/grpc/tree/aws_integration_reviewed (as discussed previously). Btw, it's completely fine to review everything in multiple PRs (in fact that's preferred) - so once you have a piece that's more or less "finished" please create a PR for it (and of course we can iterate later to improve it).
